Signal/src/network/GiphyDownloader.swift

```diff
@@ -136,41 +163,46 @@ enum GiphyAssetRequestState: UInt {
         let k10KB: UInt = 10 * 1024
         let k1KB: UInt = 1 * 1024
         for segmentSize in [k1MB, k500KB, k100KB, k50KB, k10KB, k1KB ] {
-            if fileSize >= segmentSize {
+            if contentLength >= segmentSize {
                 return segmentSize
             }
         }
-        return fileSize
+        return contentLength
     }
 
     private func createSegments() {
+        AssertIsOnMainThread()
+
         let segmentLength = segmentSize()
         guard segmentLength > 0 else {
             return
         }
-        let fileSize = UInt(contentLength)
+        let contentLength = UInt(self.contentLength)
 
         var nextSegmentStart: UInt = 0
         var index: UInt = 0
-        while nextSegmentStart < fileSize {
+        while nextSegmentStart < contentLength {
             var segmentStart: UInt = nextSegmentStart
             var redundantLength: UInt = 0
             // The last segment may overlap the penultimate segment
             // in order to keep the segment sizes uniform.
-            if segmentStart + segmentLength > fileSize {
-                redundantLength = segmentStart + segmentLength - fileSize
-                segmentStart = fileSize - segmentLength
+            if segmentStart + segmentLength > contentLength {
+                redundantLength = segmentStart + segmentLength - contentLength
```

Comment:
nit: The logic currently works out, but it seems counterintuitive that `redundantLength` will be negative for all but the last segments. I'd expect it to be 0 in those cases.

Reply:
Already reworked.  See the new arrangement.